### PR TITLE
Add changelog for 16.0.0

### LINKED ISSUE #585: Release planning for 16.0.0
JupyterHub 4 is now released and I look to get z2jh 3 and tljh 1 bundle with various new dependencies with recent breaking changes. Let's try get a release out with oauthenticator!

### Next version number: 16.0.0

We decided to go for 16.0.0 for the major refactoring.

### Things I track to resolve before release

- #594
- #615
- #620
- #599
- #601
- #602
- #603
- #605
- #628
- #630
- #631
  - Closes #619
- #637
  - Closes #632
- #640
  - Closes #639 
-  #641
  - Closes #612

### PR DESCRIPTION
**Pull request by consideRatio**
I've set the date already to tomorrow July 5, hoping that we can go for a release tomorrow, allowing z2jh / tljh to bundle with this release as well. I'm pushing onwards with this as fast as I can as I'd like to get this done before the vacation I'm taking next week for three weeks for better peace of mind. I'll not be able to drive this effort at that point, but I'm not going away entirely either.

- Closes #585 
- Closes #612

### CODE REVIEW

File: docs/source/reference/changelog.md [hunk 1]
```diff
@@ -30,34 +38,103 @@ command line for details.
...
+
+The authenticators are no longer overriding the `authenticate` method, but
+instead relying on the OAuthenticator base class `authenticate` method which
+calls a few lower level methods that can be overridden if needed. Like this, a
+lot of code has been absorbed into the OAuthenticator base class that was
+previously duplicated across authenticators.
+
+To learn more about this new structure the provider specific authenticator
+classes rely on, please for now inspect the source code for the
+`OAuthenticator.authenticate` method and `OAuthenticator.check_allowed` method.
```
Comment (manics) [inline docs/source/reference/changelog.md:68]:
```suggestion
[`OAuthenticator.authenticate`](https://github.com/jupyterhub/oauthenticator/blob/16.0.0/oauthenticator/oauth2.py#L901) and
[`OAuthenticator.check_allowed`](https://github.com/jupyterhub/oauthenticator/blob/16.0.0/oauthenticator/oauth2.py#L945) methods.
```

**Comment by consideRatio** [comment]
Rebased on main and pushed to cleanup the "files changed" view

File: docs/source/reference/changelog.md [hunk 2]
```diff
@@ -47,22 +55,86 @@ command line for details.
...
+Plans on writing more thorough documentation about this new structure is tracked
+in issue [#634](https://github.com/jupyterhub/oauthenticator/issues/634).
+
+#### New features added
+
+- breaking: add allow_existing_users config [#631](https://github.com/jupyterhub/oauthenticator/pull/631) ([@consideRatio](https://github.com/consideRatio), [@minrk](https://github.com/minrk))
```
Comment (manics) [inline docs/source/reference/changelog.md:74]: Should this be marked `[all]`?

**Review by manics — COMMENTED** [review]
Can you rebase this?

File: docs/source/reference/changelog.md (hunk 2 above)
Comment (consideRatio) [reply docs/source/reference/changelog.md:74]:
```suggestion
- [All] breaking: add allow_existing_users config [#631](https://github.com/jupyterhub/oauthenticator/pull/631) ([@consideRatio](https://github.com/consideRatio), [@minrk](https://github.com/minrk))
```

Comment (consideRatio) [reply docs/source/reference/changelog.md]: Updated the PR title as well retroactively

**Pull request merged**